openmetadata-service/src/main/java/org/openmetadata/service/security/AuthCallbackServlet.java

```diff
@@ -0,0 +1,250 @@
+import com.nimbusds.oauth2.sdk.AuthorizationCodeGrant;
+import com.nimbusds.oauth2.sdk.AuthorizationGrant;
+import com.nimbusds.oauth2.sdk.ErrorObject;
+import com.nimbusds.oauth2.sdk.ParseException;
+import com.nimbusds.oauth2.sdk.TokenErrorResponse;
+import com.nimbusds.oauth2.sdk.TokenRequest;
+import com.nimbusds.oauth2.sdk.TokenResponse;
+import com.nimbusds.oauth2.sdk.auth.ClientAuthentication;
+import com.nimbusds.oauth2.sdk.http.HTTPRequest;
+import com.nimbusds.oauth2.sdk.http.HTTPResponse;
+import com.nimbusds.oauth2.sdk.id.ClientID;
+import com.nimbusds.oauth2.sdk.id.State;
+import com.nimbusds.oauth2.sdk.pkce.CodeVerifier;
+import com.nimbusds.oauth2.sdk.token.AccessToken;
+import com.nimbusds.openid.connect.sdk.AuthenticationErrorResponse;
+import com.nimbusds.openid.connect.sdk.AuthenticationResponse;
+import com.nimbusds.openid.connect.sdk.AuthenticationResponseParser;
+import com.nimbusds.openid.connect.sdk.AuthenticationSuccessResponse;
+import com.nimbusds.openid.connect.sdk.OIDCTokenResponse;
+import com.nimbusds.openid.connect.sdk.OIDCTokenResponseParser;
+import com.nimbusds.openid.connect.sdk.op.OIDCProviderMetadata;
+import com.nimbusds.openid.connect.sdk.token.OIDCTokens;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.oidc.client.OidcClient;
+import org.pac4j.oidc.credentials.OidcCredentials;
+
+@WebServlet("/callback")
+@Slf4j
+public class AuthCallbackServlet extends HttpServlet {
+  private final OidcClient client;
+  private final ClientAuthentication clientAuthentication;
+  private final List<String> claimsOrder;
+
+  public AuthCallbackServlet(OidcClient oidcClient, List<String> claimsOrder) {
+    this.client = oidcClient;
+    this.claimsOrder = claimsOrder;
+    this.clientAuthentication = getClientAuthentication(client.getConfiguration());
+  }
+
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse resp) {
+    try {
+      String computedCallbackUrl = client.getCallbackUrl();
+      Map<String, List<String>> parameters = retrieveParameters(req);
+      AuthenticationResponse response =
+          AuthenticationResponseParser.parse(new URI(computedCallbackUrl), parameters);
+
+      if (response instanceof AuthenticationErrorResponse authenticationErrorResponse) {
+        LOG.error(
+            "Bad authentication response, error={}", authenticationErrorResponse.getErrorObject());
+        getErrorMessage(resp, new TechnicalException("Bad authentication response"));
+        return;
```

Comment:
do we need to throw exception here, lets avoid having returns in the middle of the method

Reply:
removed